rfcs/20181016-optimizer-unification.md

```diff
@@ -0,0 +1,398 @@
+
+1) `model.save()`, for a model compiled with a TF optimizer, will not include the optimizer configuration nor the optimizer state, which prevents users from restarting training from a saved model. This is due to:
+- TF Optimizer instances cannot be serialized (cloned).
+- TF Optimizer instances do not implement the Layer/Model API for weight loading/setting.
+
+2) The callbacks `LearningRateScheduler` and `ReduceLROnPlateau` (dynamic adaption of the optimizer's learning rate during training) will not work for a model compiled with a TF optimizer. This is due to the fact that there is no way to dynamically adjust the hyperparameters of a TF Optimizer after instantiating it.
+
+3) By forcing TF Optimizers for Keras training, we are asking users to take on additional complexity to use Keras. It's not enough to learn ML and NN and Keras and datasets and eager mode, now they also need to be able to know the library of TF optimizers and how to configure them. This also breaks the marketing pitch of "You can run tf.keras just like the normal keras library, with only an import change".
+
+In addition, it is fairly confusing for users to have 2 distinct sets of optimizers with a different feature set.
+
+Thus we should seek to unify the Keras optimizer API and the TensorFlow optimizer API, by 1) extending the TensorFlow optimizer API, 2) replacing the tf.keras optimizers the upgraded TF optimizers.
+
+---
+
+## Objective
+
+- Unify `tf.train` and `tf.keras.optimizers` API:
+    - Make all TensorFlow optimizers JSON-serializable, and make it possible to save/restore their state.
+    - Make it possible to dynamically modify the value of the hyperparameters of all TensorFlow optimizers, in particular the learning rate.
+        - The current way to achieve dynamic learning rates is 1) use a LR tensor with built-in decay, 2) use a callable. Both of these approaches are limited (do not support fully-dynamic rates, e.g. adapting the rate based on the current loss decrease), and not intuitive. Doing `optimizer.lr = 0.2` at arbitrary points during training is eager-first and more user-friendly.
+- Have a single set of optimizers (same signatures, same objects, no wrappers), introduced as a new set of classes with an updated API, importable from `tf.keras.optimizers`. These optimizers would be based on the existing `tf.contrib.optimizer_v2` optimizers (which themselves are based on `tf.train optimizers`).
+
+
+The old optimizers will exist in tf.compat.v1 as-is.
+
+The known breaking changes are:
+- Due to name changes, old checkpoints would not be loadable with the new optimizers. This is opt-in: your checkpoint won't break until you start using the new optimizers in your code (you can always import the old optimizers from tf.compat.v1).
+- Some arguments are getting renamed.
+- The `use_locking` argument is removed.
```

Comment:
In practice, we always keep `use_locking` as default. But whether it is going to use lock in implementation if it is removed ?